traits/trait_set_object.py

```diff
@@ -12,247 +12,425 @@
+        """ Add an element to a set.
+
+        This has no effect if the element is already present.
+
+        Parameters
+        ----------
+        value : any
+            The value to add to the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                An empty set.
+            added : set
+                A set containing the added value.
 
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitSetEvent(removed)
-                )
+        """
+        validated_values = self.validate(set([value]))
+        if len(validated_values) > 1:
+            raise ValueError("Validator returned {} values "
+                             "where 1 value is "
+                             "expected".format(len(validated_values)))
+
+        value = next(iter(validated_values))
+        if value not in self:
+            super().add(value)
+            self.notify(set(), validated_values)
 
-    def difference_update(self, value):
+    def remove(self, value):
+        """ Remove an element from a set; it must be a member.
+
+        If the element is not a member, raise a KeyError.
+
+        Parameters
+        ----------
+        value : any
+            An element in the set
+
+        Raises
+        ------
+        KeyError
+            If the value is not found in the set.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                A set containing the removed value.
+            added : set
+                An empty set.
+
+        """
+        super().remove(value)
+        self.notify(set([value]), set())
+
+    def update(self, value=()):
+        """ Update a set with the union of itself and others.
+
+        Parameters
+        ----------
+        value : iterable
+            The other iterable.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                An empty set.
+            added : set
+                A set containing the added values.
+
+        """
+        validated_values = self.validate(set(value))
+        added = validated_values.difference(self)
+
+        if len(added) > 0:
+            super().update(added)
+            self.notify(set(), added)
+
+    def difference_update(self, value=()):
+        """  Remove all elements of another set from this set.
+
+        Parameters
+        ----------
+        value : iterable
+            The other iterable.
+
+        Notes
+        -----
+        Parameters in the notification:
+            removed : set
+                A set containing the removed values.
+            added : set
+                An empty set.
+
+        """
         removed = self.intersection(value)
+
         if len(removed) > 0:
-            set.difference_update(self, removed)
+            super().difference_update(removed)
```

Comment:
I think it would be safer to try to pass on the exact same call and parameters to the superclass where possible, so call `super().difference_update(value)` unconditionally, instead of calling `super().difference_updated(removed)` conditionally. (This probably applies to other methods too; this is the method where I noticed it first.)

The idea is that we're extending the behaviour of the built-in `set`, so the code should reflect that: we're doing exactly what the superclass would do, plus some extra (validation/transformation and notification).

Reply:
Addressed in #998 .